### PR TITLE
[MM-47483] Activity Indicator while loading thread posts 

### PR DESCRIPTION
**Pull request by anurag6713**
#### Summary
1. This PR shows an activity indicator while loading thread posts instead of showing a "No replies yet" message.
2. Fixes continuous requests while loading more thread posts as the server always returns the root post. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47483

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: iOS 16, Android 11

#### Screenshots
<table>
<tr>
<td>

</td>
<td>

</td>
</tr>
</table>

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: app/utils/thread/fetch_thread_state.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+export const useFetchThreadState = () => {
+    const [state, setState] = useState(defaultState);
+    useEffect(() => {
+        const sub = subject.subscribe(setState);
+
+        return () => sub.unsubscribe();
+    }, []);
+
+    return state;
+};
```
Comment (enahum) [inline app/utils/thread/fetch_thread_state.ts:30]: this is nice, but perhaps you want to add a hook for a specific rootId ?

something like
```typescript
export const useFetchThreadStateForRoot = (rootId: string) => {
    const [isFetching, setIsFetching] = useState(false);
    useEffect(() => {
        const sub = subject.pipe(
            switchMap((s) => s[rootId] || false),
            distinctUntilChanged(),
        ).subscribe(setIsFetching)

        return () => sub.unsubscribe();
    }, []);

    return state;
};
```

Comment (enahum) [reply app/utils/thread/fetch_thread_state.ts:30]: Also, this one would make sense to move to the `hooks` folder while the Behavior can be moved to the `store` folder? that way is easy to find and maintain?

@larkox what do you think if we move the fetching posts event to something like this? I think it'll be less racy

File: app/actions/remote/post.ts [hunk 2]
```diff
@@ -27,6 +27,7 @@ import {getValidEmojis, matchEmoticons} from '@utils/emoji/helpers';
 import {logError} from '@utils/log';
 import {processPostsFetched} from '@utils/post';
 import {getPostIdsForCombinedUserActivityPost} from '@utils/post_list';
+import {setFetchThreadState} from '@utils/thread/fetch_thread_state';
```
Comment (enahum) [inline app/actions/remote/post.ts:30]: see the comment about the path

File: app/components/post_list/thread_overview/thread_overview.tsx [hunk 3]
```diff
@@ -98,30 +100,44 @@ const ThreadOverview = ({isSaved, repliesCount, rootPost, style, testID}: Props)
...
+        repliesCountElement = (
+            <FormattedText
+                style={styles.repliesCount}
+                id='thread.repliesCount'
+                defaultMessage='{repliesCount, number} {repliesCount, plural, one {reply} other {replies}}'
+                testID={`${testID}.replies_count`}
+                values={{repliesCount}}
+            />
+        );
+    } else if (rootPost?.id && fetchThreadState[rootPost.id]) {
```
Comment (enahum) [inline app/components/post_list/thread_overview/thread_overview.tsx:114]: with the hook that uses the rootId this is simplified

File: app/screens/thread/thread_post_list/thread_post_list.tsx [hunk 4]
```diff
@@ -42,21 +45,29 @@ const ThreadPostList = ({
 }: Props) => {
     const isTablet = useIsTablet();
     const serverUrl = useServerUrl();
+    const theme = useTheme();
+    const fetchThreadState = useFetchThreadState();
 
-    const canLoadPosts = useRef(true);
+    const canLoadMorePosts = useRef(true);
     const fetchingPosts = useRef(false);
```
Comment (enahum) [inline app/screens/thread/thread_post_list/thread_post_list.tsx:52]: fetchingPosts should be true if fetchThread for the post is true, in fact do you need both?

File: app/screens/thread/thread_post_list/thread_post_list.tsx [hunk 5]
```diff
@@ -95,6 +111,7 @@ const ThreadPostList = ({
             rootId={rootPost.id}
             shouldShowJoinLeaveMessages={false}
             showMoreMessages={isCRTEnabled}
+            header={header}
```
Comment (enahum) [inline app/screens/thread/thread_post_list/thread_post_list.tsx:114]: if the thread is being fetched, perhaps we need to prevent the pull to refresh functionality ?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/screens/thread/thread_post_list/thread_post_list.tsx (hunk 4 above)
Comment (anurag6713) [reply app/screens/thread/thread_post_list/thread_post_list.tsx:52]: True, we can reuse the new hook for this condition as well.

File: app/client/rest/constants.ts [hunk 6]
```diff
@@ -12,4 +12,4 @@ export const HEADER_X_CSRF_TOKEN = 'X-CSRF-Token';
 export const HEADER_X_VERSION_ID = 'X-Version-Id';
 export const DEFAULT_LIMIT_BEFORE = 30;
 export const DEFAULT_LIMIT_AFTER = 30;
-export const PER_PAGE_DEFAULT = 60;
+export const PER_PAGE_DEFAULT = 30;
```
Comment (enahum) [inline app/client/rest/constants.ts:15]: what is the reason behind this change?

File: app/components/post_list/post_list.tsx [hunk 7]
```diff
@@ -145,6 +149,9 @@ const PostList = ({
     }, []);
 
     const onRefresh = useCallback(async () => {
+        if (disablePullToRefresh) {
+            return;
+        }
```
Comment (enahum) [inline app/components/post_list/post_list.tsx:154]: at this point is late, cause the callback is being executed, I do thing that you need to change the `enabled` prop of the `PostListRefreshControl` component in line 375

**Review by enahum — CHANGES_REQUESTED** [review]
one more change and it should be gtg

File: app/utils/thread/fetch_thread_state.ts (hunk 1 above)
Comment (larkox) [reply app/utils/thread/fetch_thread_state.ts]: Sure! @anurag6713 This has been my approach, which is similar to other similar subjects in the managers.
https://github.com/mattermost/mattermost-mobile/pull/6872
If you want, we can sync to see how we can make it more similar.

File: app/store/fetching_thread_store.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+type State = {[id: string]: boolean};
+
+const defaultState: State = {};
+
+export const subject: BehaviorSubject<State> = new BehaviorSubject(defaultState);
+
+export const setFetchingThreadState = (rootId: string, isFetching: boolean) => {
+    const prevState = subject.value;
+    subject.next({
+        ...prevState,
```
Comment (larkox) [inline app/store/fetching_thread_store.ts:15]: This will grow the object in an unbounded way. Probably not a big deal, but if we travel through many threads without closing the app this may become big. 0/5 whether you should delete the element from the state when `isFetching` is false.

File: app/store/fetching_thread_store.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        [rootId]: isFetching,
```
Comment (larkox) [inline app/store/fetching_thread_store.ts:16]: Probably not a big deal, but there is a remote possibility that the same rootId happens in different servers. 0/5 on whether it makes sense to divide this by serverUrl.

File: app/screens/thread/thread_post_list/thread_post_list.tsx [hunk 10]
```diff
@@ -42,23 +45,28 @@ const ThreadPostList = ({
...
             if (lastPost) {
                 options.fromPost = lastPost.id;
                 options.fromCreateAt = lastPost.createAt;
             }
             const result = await fetchPostThread(serverUrl, rootPost.id, options);
-            fetchingPosts.current = false;
-            canLoadPosts.current = Boolean(result?.posts?.length);
+
+            // Root post is always fetched, so the result would include +1
+            canLoadMorePosts.current = (result?.posts?.length || 0) >= PER_PAGE_DEFAULT;
```
Comment (larkox) [inline app/screens/thread/thread_post_list/thread_post_list.tsx:65]: If always include X+1, shouldn't we check for just `>` instead of `>=`?

**Review by larkox — APPROVED** [review]
LGTM, just a few minor comments.

**Comment by larkox** [comment]
@anurag6713 You and I have approached the subscription in two different ways.
- (Me) Return the subject as an observable and use the withObservables HOC
- (You) Subscribe directly to it in a effect with a setState

I don't have a strong preference about this, but since they seem "two different ways of doing the same", we should decide on one. Do you have any opinion on which one is better?

File: app/store/fetching_thread_store.ts (hunk 9 above)
Comment (enahum) [reply app/store/fetching_thread_store.ts:16]: I don't think that is needed, unless until now, there is no way to see a thread from a server different than the current server

**Comment by enahum** [comment]
> @anurag6713 You and I have approached the subscription in two different ways.
> 
> * (Me) Return the subject as an observable and use the withObservables HOC
> * (You) Subscribe directly to it in a effect with a setState
> 
> I don't have a strong preference about this, but since they seem "two different ways of doing the same", we should decide on one. Do you have any opinion on which one is better?

@larkox I did follow this approach for the android share extension, by using hooks instead of the `withObservables` HOC, I just wanted to share that here just in case

**Review by enahum — APPROVED** [review]
